src/output-json-ftp.c

```diff
@@ -0,0 +1,315 @@
+
+#include "util-unittest.h"
+#include "util-buffer.h"
+#include "util-debug.h"
+#include "util-mem.h"
+#include "util-byte.h"
+
+#include "output.h"
+#include "output-json.h"
+
+#include "app-layer.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-ftp.h"
+#include "output-json-ftp.h"
+
+#ifdef HAVE_LIBJANSSON
+
+typedef struct LogFTPFileCtx_ {
+    LogFileCtx *file_ctx;
+    OutputJsonCommonSettings cfg;
+} LogFTPFileCtx;
+
+typedef struct LogFTPLogThread_ {
+    LogFTPFileCtx *ftplog_ctx;
+    uint32_t            count;
+    MemBuffer          *buffer;
+} LogFTPLogThread;
+
+/*
+ * \brief Returns the ending offset of the next line.
+ *
+ * Here, "next line" is defined as terminating on
+ * - Newline character
+ * - Null character
+ *
+ * \param buffer Contains zero or more characters.
+ * \param len Size, in bytes, of buffer.
+ *
+ * \retval Offset from the start of buffer indicating the where the
+ * next "line ends". The characters between the input buffer and this
+ * value comprise the line.
+ *
+ * NULL is found first or a newline isn't found, then
+ */
+static uint16_t JsonGetNextLineFromBuffer(const char *buffer, const uint16_t len)
+{
+    if (!buffer || *buffer == '\0')
+        return UINT16_MAX;
+
+    char *c = strchr(buffer, '\n');
+    return c == NULL ? len : c - buffer + 1;
+}
+
+static json_t *JsonFTPLogCommand(Flow *f, FTPTransaction *tx)
+{
+    json_t *cjs = json_object();
+    if (!cjs) {
+        return cjs;
+    }
+
+    /* Preallocate array objects to simplify failure case */
+    json_t *js_resplist;
+    json_t *js_respcode_list;
+    if (!TAILQ_EMPTY(&tx->response_list)) {
+        js_resplist = json_array();
+        js_respcode_list = json_array();
+
+        if (unlikely(js_resplist == NULL || js_respcode_list == NULL)) {
+            if (js_resplist) {
+                json_decref(js_resplist);
+            } else {
+                json_decref(js_respcode_list);
+            }
+            return cjs;
+        }
+    }
+
+    json_object_set_new(cjs, "command",
+                        json_string(tx->command_descriptor->command_name_upper));
+    uint32_t min_length = tx->command_descriptor->command_length + 1; /* command + space */
+    if (tx->request_length > min_length) {
+        json_object_set_new(cjs, "command_data",
+                            JsonAddStringN((const char *)tx->request + min_length,
+                                           tx->request_length - min_length));
+    } else {
+        json_object_set_new(cjs, "command_data", json_string(NULL));
+    }
+
+    if (!TAILQ_EMPTY(&tx->response_list)) {
+        FTPString *response;
+        TAILQ_FOREACH(response, &tx->response_list, next) {
+            /* handle multiple lines within the response, \r\n delimited */
+            uint8_t *where = response->str;
+            uint16_t length = response->len;
+            uint16_t pos;
+            while ((pos = JsonGetNextLineFromBuffer((const char *)where, length)) != UINT16_MAX) {
```

Comment:
Should this parsing rather be done in app-layer-ftp.c ?

Reply:
IMO, the parsing is best deferred until needed. If FTP logging is not enabled, then the parsing is not needed.

Reply:
Ok. but then could output-json-ftp.c call some new parsing function in app-layer-ftp.c ?
This way, we will defer parsing until needed and the parsing will be in app-*.c files